src/waiter.c

```diff
@@ -0,0 +1,311 @@
+        waiter->callbacks_called = true;
+        sm_fault(waiter->sm);
+        sm_exec_end(waiter->sm);
+    }
+    srw_lock_release_exclusive(waiter->lock);
+}
+
+static void dispose(void* context)
+{
+    (void)context;
+}
+
+static void try_complete(WAITER_HANDLE waiter)
+{
+    if (waiter->registered && waiter->notified && !waiter->callbacks_called)
+    {
+        sm_fault(waiter->sm);
+        waiter->callbacks_called = true;
+        sm_exec_end(waiter->sm);
+        waiter->notification_callback(waiter->notification_callback_context, waiter->data, WAITER_RESULT_OK);
+        THANDLE_ASSIGN(RC_PTR)(&waiter->data, NULL);
+        sm_exec_end(waiter->sm);
+        waiter->notify_complete_callback(waiter->notify_complete_callback_context, WAITER_RESULT_OK);
+    }
+}
+
+THANDLE(ASYNC_OP) waiter_register_notification(WAITER_HANDLE waiter, NOTIFICATION_CALLBACK notification_callback, void* context)
+{
+    THANDLE(ASYNC_OP) result = NULL;
+    if (waiter == NULL || notification_callback == NULL || context == NULL)
+    {
+        LogError("Invalid arguments: WAITER_HANDLE waiter = %p, NOTIFICATION_CALLBACK notification_callback = %p, void* context = %p",
+            waiter, notification_callback, context);
+    }
+    else
+    {
+        SM_RESULT sm_exec_begin_result = sm_exec_begin(waiter->sm);
+        if (sm_exec_begin_result != SM_EXEC_GRANTED)
+        {
+            LogError("Failure in sm_exec_begin(waiter->sm=%p)", waiter->sm);
+        }
+        else
+        {
+            bool success;
+            srw_lock_acquire_exclusive(waiter->lock);
+            {
+                if (waiter->registered)
+                {
+                    LogError("waiter_register_notification called twice on WAITER waiter=%p."
+                    " Previous arguments: NOTIFICATION_CALLBACK notification_callback=%p, void* context=%p"
+                    " Current arguments: NOTIFICATION_CALLBACK notification_callback=%p, void* context=%p",
+                    waiter,
+                    waiter->notification_callback, waiter->notification_callback_context,
+                    notification_callback, context
+                    );
+                    success = false;
+                }
+                else
+                {
+                    THANDLE(ASYNC_OP) async_op = async_op_create(cancel_register_notification, sizeof(WAITER_HANDLE), alignof(WAITER_HANDLE), dispose);
+                    if (async_op == NULL)
+                    {
+                        LogError("Failure in async_op_create(cancel_register_notification=%p, sizeof(WAITER_HANDLE)=%" PRIu32 ", alignof(WAITER_HANDLE)=%" PRIu32 ", dispose=%p)", cancel_register_notification, (uint32_t)sizeof(WAITER_HANDLE), (uint32_t)alignof(WAITER_HANDLE), dispose);
+                        success = false;
+                    }
+                    else
+                    {
+                        ((ASYNC_OP*)async_op)->context = waiter;
```

Comment:
>(ASYNC_OP*) [](http://example.com/codeflow?start=25&length=11)

is this cast necessary?

Reply:
Yes, I get "error C2166: l-value specifies const object" otherwise